Add MinIdleConnsHeadroom support

pracucci opened the pull request:
Currently memcached client allows to configure the max number of idle connections. These idle connections are never closed, even if unused for a long time. There's some tensions on the value you set for idle connections:
- Too low: new connections are opened for each Get operation when your concurrency is > max idle connections
- Too high: a short spike in Get concurrency will open new connections which then are kept idle until the process restarts even if unused

In this PR I'm proposing to add a `MinIdleConnsHeadroom` config option. When configured, the client periodically closes idle connections up until `MinIdleConnsHeadroom` percentage of the recently used connections. If `MinIdleConnsHeadroom` is not configured (zero value) there's no difference compared to previous version.

In order to periodically close the connections, the client now runs a maintenance goroutine. The client goroutine can be stopped calling the new `Close()` function.

_Other the new unit tests, I've manually tested it in Mimir (see https://github.com/grafana/mimir/pull/4249) and looks working as designed._

colega commented on memcache/memcache.go:
```diff
@@ -238,13 +270,86 @@ func (c *Client) netTimeout() time.Duration {
+	defer c.lk.Unlock()
+
+	// Nothing to do if there are no free connections.
+	if c.freeconn == nil {
```
Nit: we don't need this, the loop will do the same.

colega commented:
I would consider an alternative logic: define the `MinIdleConnsHeadroom` as the percentage of conns using the last specified period.

Let me explain myself: consider a 1 minute period, if a connection has been used during that last 1 minute period, then we don't consider it "idle" at all: it is idle in terms of the network, but we know that some load spike during the last minute has used it, so we shouldn't really close it. Let's call these connections used during the last 1 minute "recently used".

Let's say we have 100 active connections, then with `MinIdleConnsHeadroom=50` we should leave 50 really-idle connections on top of those 100 recently used, to account for the possible spikes.

If the use decreases, let's say to 50 active conns, then we'll close connections until we have 75 (50 recently used + 25 idle). This will not happen in a single close, as the number of active connections will naturally decrease over time.

I see two advantages in this:
- The number of idle connections we keep depends on the number of used ones, rather than an absolute minimum.
- The logic is (IMO) simpler to implement: take the slice of conns, count the recently used ones, and then skip the percentage of idle we want to keep, and close the rest.

---

This is just an alternative idea to consider, I'm not saying there's something wrong with your logic 👍

pracucci commented:
> I would consider an alternative logic: define the MinIdleConnsHeadroom as the percentage of conns using the last specified period.

I like your idea because you don't have to think about the absolute minimum value to set. I'm going to change the implementation.

colega commented on memcache/memcache.go:
```diff
@@ -238,13 +278,91 @@ func (c *Client) netTimeout() time.Duration {
+		// 2. Keep at least 1 idle connection.
+		numRecentlyUsed := len(freeConnections) - numIdle
+		numIdleToKeep := int(math.Max(1, math.Ceil(float64(numRecentlyUsed)*minIdleHeadroom)))
+		numIdleToClose := len(freeConnections) - numRecentlyUsed - numIdleToKeep
```
Maybe this is easier to follow? Not a strong opinion:

```suggestion
		numIdleToClose := numIdle - numIdleToKeep
```

pracucci replied:
👍 Done in https://github.com/grafana/gomemcache/pull/10/commits/c873d2f48a2fa56f66e751f71e6a9003f208c702.

colega reviewed (approved):
LGTM, thanks!

colega commented on memcache/memcache.go:
```diff
@@ -238,13 +278,91 @@ func (c *Client) netTimeout() time.Duration {
+
+		// Compute the number of connections to close, honoring two properties:
+		// 1. Keep a number of idle connections equal to the configured headroom.
+		// 2. Keep at least 1 idle connection.
```
Why do we need to keep at least 1 idle connection? You already have the number of recently used ones, you may not want to have any extra ones.

pracucci replied:
Agree. Done in https://github.com/grafana/gomemcache/pull/10/commits/4413597d09e84fe8e25f9d5941d591e57ef338df.

colega commented:
Left some opinionated comments, they're not blockers, but consider thinking about them.

pracucci commented on memcache/memcache.go:
```diff
@@ -238,13 +278,91 @@ func (c *Client) netTimeout() time.Duration {
 }
 
+func (c *Client) minIdleConnsHeadroom() float64 {
+	if c.MinIdleConnsHeadroom < 0 || c.MinIdleConnsHeadroom >= 1 {
```
Reason was that to me didn't make much sense, but doesn't hurt to support it. Done in https://github.com/grafana/gomemcache/pull/10/commits/4413597d09e84fe8e25f9d5941d591e57ef338df.

pracucci commented:
Thanks @colega for your review. I should have addressed all comments, plus added few more test cases and made `Close()` idempotent. Can you take another look, please?

colega reviewed (approved):
Looks awesome! Thank you!
